Plot probe

hargunkaur286 opened the pull request:
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/neuroinformatics-unit/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [ ] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

**What does this PR do?**

## References

Please reference any existing issues/PRs that relate to this PR.

## How has this PR been tested?

Please explain how any new code has been tested, and how you have ensured that no existing functionality has changed.

## Is this a breaking change?

If this PR breaks any existing functionality, please explain how and why.

## Does this PR require an update to the documentation?

If any features have changed, or have been added. Please explain how the
documentation has been updated.

## Checklist:

- [ ] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

adamltyson commented:
Thanks for this @hargunkaur286, just FYI due to the main developer of spikewrap being away, it may be some time until this can be reviewed.

hargunkaur286 commented:
> Thanks for this @hargunkaur286, just FYI due to the main developer of spikewrap being away, it may be some time until this can be reviewed.

Hi @adamltyson,
Thank you for the update. I saw the note on the previous PR (#231) mentioning that @JoeZiminski is away, so I completely understand. I’ll wait patiently for the review. Thanks again!

JoeZiminski commented on spikewrap/structure/session.py:
```diff
@@ -668,6 +668,49 @@ def _get_concat_raw_run(self) -> ConcatRawRun:
+        Returns:
+            The Matplotlib figure containing the probe plot, or None if no probe is available.
+        """
+        if not self.runs:
```
Instead can raise an error here

JoeZiminski commented on spikewrap/structure/session.py:
```diff
@@ -668,6 +668,49 @@ def _get_concat_raw_run(self) -> ConcatRawRun:
+            _utils.message_user("No runs available in this session.")
+            return None
+
+        first_run = self.runs[0]
```
This section is good, great idea to check all probes match. Following on from the suggestion above, it would be nice if the run had a get_probe().

Then you can do something like:
```
for run in  self._pp_runs:
... run.get_probe()... check equality against first run probe
```

and then handle the probe plotting / saving here.

There will be two cases, either the probe is grouped shank or split per shank. If split per shank, then we will need to get the probe for each shank.

For plotting, if multi-shank this can be on a single plot with subplots.

JoeZiminski reviewed (changes requested):
Hey @hargunkaur286 thanks a lot for this! Very nice approach, please see suggestions mostly around refactoring.

hargunkaur286 commented:
Hello @JoeZiminski , 
I've implemented the requested updates related to probe plotting and integration. 

I added a get_probe() method to the PreprocessedRun class, which returns a dictionary of probe objects indexed by shank ID (e.g. "grouped", "shank_0", etc.) like you suggested. Also, I moved the probe plotting logic to a new method Session.plot_probe() to ensure that all runs have consistent probe structures before plotting and it handles both grouped and per-shank cases.

JoeZiminski commented:
Hey @hargunkaur286 thanks a lot for this, this is looking good thanks also for the tests. Currently I'm on paternity leave so will not have time to review and test this in full prior to the GSoC deadline, of course do include this contribution in your application nonetheless.

Some quick feedback, I would suggest removing all plotting functionality from the `_preprocessing_run` class and only expose a simple getter. Plotting / saving can be handled entirely at the session level (we are unlikely to ever save a separate probe plot for each run, as they should be the same for all runs within a session. Cheers!

hargunkaur286 commented:
> Hey @hargunkaur286 thanks a lot for this, this is looking good thanks also for the tests. Currently I'm on paternity leave so will not have time to review and test this in full prior to the GSoC deadline, of course do include this contribution in your application nonetheless.
> 
> Some quick feedback, I would suggest removing all plotting functionality from the `_preprocessing_run` class and only expose a simple getter. Plotting / saving can be handled entirely at the session level (we are unlikely to ever save a separate probe plot for each run, as they should be the same for all runs within a session. Cheers!

Thanks a lot for the quick feedback, really appreciate it! I've updated the implementation to remove plotting logic from _preprocess_run and now handle all probe plotting/saving at the session level as suggested. Thank you again, and enjoy your paternity leave! Meanwhile, I'll try to solve other open issues :)
